utility_analysis_new/parameter_tuning.py

```diff
@@ -0,0 +1,82 @@
+import pipeline_dp
+from dataclasses import dataclass
+import operator
+
+
+@dataclass
+class FrequencyBin:
+    """Represents 1 bin of the histogram.
+
+    The bin represents integers between 'lower' (inclusive) and 'upper'
+     (exclusive, not stored in this class).
+
+    Attributes:
+        lower: the lower bound of the bin.
+        count: the number of elements in the bin.
+        sum: the sum of elements in the bin.
+        max: the maximum element in the bin (which might be smaller than upper-1).
+    """
+    lower: int
+    count: int
+    sum: int
+    max: int
+
+    def __add__(self, other: 'FrequencyBin') -> 'FrequencyBin':
+        return FrequencyBin(self.lower, self.count + other.count,
+                            self.sum + other.sum, max(self.max, other.max))
+
+    def __eq__(self, other):
+        return self.lower == other.lower and self.count == other.count and self.sum == other.sum and self.max == other.max
+
+
+def _to_bin_lower(n: int) -> int:
+    """Finds the lower bound of the histogram bin which contains ."""
+    # For scalability reasons bins can not be all width=1. For the goals of
+    # contribution computations it is ok to make bins of large values, to be
+    # larger.
+    # Here, the following strategy is used: n is rounded down, such that only 3
+    # left-most digits of n is kept, e.g. 123->123, 1234->1230, 12345->12300.
+    bound = 1000
+    while n > bound:
+        bound *= 10
+
+    round_base = bound // 1000
+    return n // round_base * round_base
+
+
+def _compute_frequency_histogram(
+        col, backend: pipeline_dp.pipeline_backend.PipelineBackend):
+    """Computes histogram of elements frequencies in collection.
+
+    Args:
+        col: collection with positive integers.
+        backend: PipelineBackend to run operations on the collection.
+    Returns:
+        1 element collection, which contains a list of FrequencyBin sorted by
+        'lower' attribute.
+    """
+
+    col = backend.count_per_element(col, "Frequency of elements")
+
+    # Combiner elements to histogram buckets of increasing sizes. Having buckets
+    # of size = 1 is not scalable.
+    col = backend.map_tuple(
+        col, lambda n, f:
+        (_to_bin_lower(n),
+         FrequencyBin(lower=_to_bin_lower(n), count=f, sum=f * n, max=n)),
+        "To FrequencyBin")
+
+    # (lower_bin_value, FrequencyBin)
```

Comment:
We can't have this comment line be under the previous code line again? Maybe having a nextline after this comment helps?